Source/WebCore/loader/ImageLoader.cpp

```diff
@@ -206,106 +211,160 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
+            // In this case, we should behave same as attr.isNull().
+            imageURL = imageElement->currentURL();
+            if (!imageURL.isValid()) {
+                didUpdateCachedImage(relevantMutation, WTFMove(newImage));
+                return;
+            }
+        } else
+            imageURL = document->completeURL(attr);
+        m_pendingURL = attr;
+    }
+    ResourceRequest resourceRequest(imageURL);
+    resourceRequest.setInspectorInitiatorNodeIdentifier(InspectorInstrumentation::identifierForNode(m_element));
+
+    auto request = createPotentialAccessControlRequest(WTFMove(resourceRequest), WTFMove(options), document, crossOriginAttribute);
+    request.setInitiator(element());
+    m_imageComplete = false;
+
+    // https://html.spec.whatwg.org/C#update-the-image-data step 7.4
+    if (!imageElement || imageElement->document().isImageDocument() || (!imageElement->usesSrcsetOrPicture() && canReuseFromListOfAvailableImages(request, document))) {
+        doUpdateFromElement(relevantMutation, WTFMove(request));
+        return;
+    }
+    if (m_microtaskQueued)
+        return;
+
+    if (!document->getCodePosition()) {
+        // Grab the source location here, from the stack. Then store it on the document and extract it on the other end.
+        // This is needed to ensure CSP reports get the right source location, regardless of async image loading.
+        auto codePosition = ContentSecurityPolicy::getCurrentCodePosition();
+        if (codePosition)
+            document->setCodePosition(codePosition.value());
+    }
+    m_microtaskQueued = true;
+    // https://html.spec.whatwg.org/C#update-the-image-data step 8
+    document->eventLoop().queueMicrotask([imageLoader = this, relevantMutation, moved_request = WTFMove(request)] mutable {
+        imageLoader->clearMicrotaskQueuedBool();
+        imageLoader->doUpdateFromElement(relevantMutation, WTFMove(moved_request));
+    });
+}
+
+void ImageLoader::doUpdateFromElement(RelevantMutation relevantMutation, CachedResourceRequest&& request)
+{
+    Ref document = element().document();
+    RefPtr imageElement = dynamicDowncast<HTMLImageElement>(element());
+    AtomString crossOriginAttribute;
+    if (imageElement) {
+        URL imageURL = imageElement->currentURL();
+        if (request.resourceRequest().url() != imageURL) {
+            ResourceLoaderOptions options = CachedResourceLoader::defaultCachedResourceOptions();
+            options.contentSecurityPolicyImposition = protectedElement()->isInUserAgentShadowTree() ? ContentSecurityPolicyImposition::SkipPolicyCheck : ContentSecurityPolicyImposition::DoPolicyCheck;
+            options.loadedFromPluginElement = is<HTMLPlugInElement>(element()) ? LoadedFromPluginElement::Yes : LoadedFromPluginElement::No;
+            options.sameOriginDataURLFlag = SameOriginDataURLFlag::Set;
+            options.serviceWorkersMode = is<HTMLPlugInElement>(element()) ? ServiceWorkersMode::None : ServiceWorkersMode::All;
```

Comment:
This is part is repeated in `updateFromElement()`. Can we remove this duplication?